### PR TITLE
Restrict dtype of arguments in some of unary operators in basic_math

### PR DESCRIPTION
**Pull request by delta2323**
The objective of this fix is basically same as d7836e .

Some unary operators in `basic_math` is implemented by explicitly writing kernel code. In that case, dtypes of arguments are restricted to float32. This PR imposes this restriction to `check_type_forward` of the functions.

Related to #219

### CODE REVIEW

**Comment by beam2d** [comment]
Looks good. How about the dtypes of following functions?
- AddConstant
- SubFromConstant
- MulConstant
- Exp
- Log

**Comment by delta2323** [comment]
The followings are my understandings.
As for `Exp` and `Log` it is enough that their arguments are `float`. We do not restrict its size.

For `***Constat`(take `Addconstant` which maps `x` to `x+a` for example)
- `a` is scalar :  It is no problem if the constant `a` is a scalar because of `_force_type`.
- `a` is array
    - `x` is int : We do not admit this case (as we do in other functions)
    - `x` is float64 : Promotion causes no problems
    - `x` is float32
        - `a` is int : Type check should be passed. We should coarse `a` to float32.
        - `a` is float32 : No problem
        - `a` is float64 : Type check should forbids the case.

**Comment by beam2d** [comment]
Thank you for the investigation. How about forbidding an int array `a`? I feel inconsistency on prohibiting a float64 array while allowing an int array (by forcing the dtype).

**Comment by delta2323** [comment]
The other alternative to handle the case float64 is to coarse it to float32.

**Comment by delta2323** [comment]
In the end,  `Exp` and `Log` restrict the kind of input arguments to `float` and `***Constant` forbids to use `list` or `tuple` as a constant (use `numpy.ndarray` instead)

**Comment by unnonouno** [comment]
Do we need to test GPUarray?

**Comment by unnonouno** [comment]
How about `sin` and `cos`?

File: tests/functions_tests/test_basic_math.py [hunk 1]
```diff
@@ -223,62 +223,232 @@ def make_data(self):
...
+
+        self._test_constant_one(func, x_data, 1)
+        self._test_constant_one(func, x_data, 1.0)
+        self._test_constant_one(func, x_data, numpy.int64(1))
+        self._test_constant_one(func, x_data, numpy.float64(1.0))
+
+    def _test_constant_gpu(self, func):
+        x_data = numpy.array(1, numpy.float32)
+
+        self._test_constant_one(func, x_data, 1)
```
Comment (unnonouno) [inline tests/functions_tests/test_basic_math.py]: `gpu=True`?

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**